velox/common/base/Semaphore.h

```diff
@@ -16,55 +16,33 @@
 
 #pragma once
 
-#include <condition_variable>
-#include <cstdint>
-#include <functional>
-#include <memory>
 #include <mutex>
-
-#include "velox/common/base/Exceptions.h"
-#include "velox/common/future/VeloxPromise.h"
+#include <semaphore>
 
 namespace facebook::velox {
 
-/// Similar to std::counting_semaphore in C++ 20.
+/// Wrapper with mutex around std::counting_semaphore in C++ 20.
+/// std::counting_semaphore is not bound to a thread but the mutex is.
 class Semaphore {
  public:
-  Semaphore(int32_t count) : count_(count) {}
+  Semaphore(int32_t count) : sem_(count) {}
 
   // If count > 0, decrements count. Otherwise blocks until count is incremented
   // by release() from another thread.
   void acquire() {
     std::unique_lock l(mutex_);
-    if (count_) {
-      --count_;
-    } else {
-      ++numWaiting_;
-      cv_.wait(l, [&]() { return count_ > 0; });
-      --numWaiting_;
-      --count_;
-    }
+    sem_.acquire();
```

Comment:
I kept the previous behavior. There is a difference between the counting_semaphore and the mutex in how they work (what they protect - a semaphore can be changed by other threads, with the mutex this is prevented). 
And so I used the counting_semaphore to replace the volatile counter (and the way it was triggered). 
Because the mutex is used before (with its behavior) I didn't want to change the behavior and allow new race conditions. This can be revisited as a TODO.

Reply:
Then you probably should not change it, and add the reason to comment that this cannot be a drop-in replacement for `std:: counting_semaphore`.  The change now will cause deadlock.

Reply:
I can undo it and remove the volatile keyword for the members - which is what the complaint of the compiler is.

When I made the change I thought about the behavior change. The fact that a mutex was always taken hasn't changed. I understood that using the counting_semaphore is a bit more permissive in terms of access. The fact that the mutex is taken still allows only a single thread to access it at any given time. But it doesn't change the counting.

Reply:
A thread blocked on `acquire` would hold the lock and block another thread to call `release`. 
 `condition_variable::wait` is releasing the lock in the current code, that's a behavior change.